### PR TITLE
chore: validate ROCM support

### PR DESCRIPTION
**Pull request by phillip-gaisford**
## Description
ROCM GPUs are available on the Grenoble machine (o184i023). Make some changes to pass necessary
variables from the Slurm to the Determined environment.

## Test Plan
Run the master:
`./tools/slurmcluster.sh -A o184i054   
`

Run the agent:
`gaisford@o184i023:~$ srun -p mlde_rocm --gpus-per-node=1 /cstor/determined/agent/determined-agent --master-host=o184i054.gre.smktg.hpecorp.net  --master-port=8086 --resource-pool=default --container-runtime=singularity --slot-type=rocm
`

Configure the experiment:

```
(dai-ee) phillipgaisford@16 cifar10_pytorch % more const-rocm.yaml
name: cifar10_pytorch_const
debug: true
max_restarts: 0
environment:
  image:  /cstor/determined/images/determinedai/environments:rocm-5.5-pytorch-1.10-tf-2.10-rocm-0.21.2.1687535582
...
```

Log snippet:

```
...
[88c6cd33] singularity run \ 	--writable-tmpfs \
	--pwd /run/determined/workdir \ 	--env-file /tmp/determined-gaisford-o186i122/706594684-88c6cd33-22e7-4c67-9ac8-4074e5c6a35a/envfile \
	--bind /tmp/determined-gaisford-o186i122/706594684-88c6cd33-22e7-4c67-9ac8-4074e5c6a35a/archives/etc/ssh/ssh_config:/etc/ssh/ssh_config \ 	--bind /tmp/determined-gaisford-o186i122/706594684-88c6cd33-22e7-4c67-9ac8-4074e5c6a35a/archives/opt/determined:/opt/determined \
	--bind /tmp/determined-gaisford-o186i122/706594684-88c6cd33-22e7-4c67-9ac8-4074e5c6a35a/archives/run/determined:/run/determined \ 	--bind /etc/hosts:/etc/hosts \
	--bind /cstor/determined/checkpoints:/determined_shared_fs \ 	--rocm /cstor/determined/images/determinedai/environments:rocm-5.5-pytorch-1.10-tf-2.10-rocm-0.21.2.1687535582 /run/determined/singularity-entrypoint-wrapper.sh /run/determined/train/entrypoint.sh
<info> [2023-06-30 18:35:12] [88c6cd33] WARNING: Overriding HOME environment variable with APPTAINERENV_HOME is not permitted
<info> [2023-06-30 18:35:13] [88c6cd33] /run/determined/singularity-entrypoint-wrapper.sh: line 10: [: =: unary operator expected
<info> [2023-06-30 18:35:13] [88c6cd33] INFO: executing /run/determined/train/entrypoint.sh
<info> [2023-06-30 18:35:20] [88c6cd33] [7074] root: detected 0 gpus (nvidia-smi not found)
<info> [2023-06-30 18:35:21] [88c6cd33] [7089] root: New trial runner in (container 88c6cd33-22e7-4c67-9ac8-4074e5c6a35a) on agent o186i122: {"bind_mounts": [{"host_path": "/etc/hosts", "container_path": "/etc/hosts", "read_only": false, "propagation": "rprivate"}], "checkpoint_policy": "best", "checkpoint_storage": {"host_path": "/cstor/determined/checkpoints", "propagation": "rprivate", "save_experiment_best": 0, "save_trial_best": 1, "save_trial_latest": 1, "storage_path": null, "type": "shared_fs"}, "data": {}, "debug": true, "description": null, "entrypoint": "model_def:CIFARTrial", "environment": {"image": {"cpu": "/cstor/determined/images/determinedai/environments:rocm-5.5-pytorch-1.10-tf-2.10-rocm-0.21.2.1687535582", "cuda": "/cstor/determined/images/determinedai/environments:rocm-5.5-pytorch-1.10-tf-2.10-rocm-0.21.2.1687535582", "rocm": "/cstor/determined/images/determinedai/environments:rocm-5.5-pytorch-1.10-tf-2.10-rocm-0.21.2.1687535582"}, "environment_variables": {"cpu": ["NCCL_DEBUG=INFO", "NCCL_SOCKET_IFNAME=ens,eth,ib"], "cuda": ["NCCL_DEBUG=INFO", "NCCL_SOCKET_IFNAME=ens,eth,ib"], "rocm": ["NCCL_DEBUG=INFO", "NCCL_SOCKET_IFNAME=ens,eth,ib"]}, "proxy_ports": [], "ports": {"trial": 1734}, "registry_auth": null, "force_pull_image": false, "pod_spec": null, "add_capabilities": [], "drop_capabilities": []}, "hyperparameters": {"global_batch_size": {"type": "const", "val": 32}, "layer1_dropout": {"type": "const", "val": 0.25}, "layer2_dropout": {"type": "const", "val": 0.25}, "layer3_dropout": {"type": "const", "val": 0.5}, "learning_rate": {"type": "const", "val": 0.0001}, "learning_rate_decay": {"type": "const", "val": 1e-06}}, "labels": [], "max_restarts": 0, "min_checkpoint_period": {"batches": 0}, "min_validation_period": {"epochs": 1}, "name": "cifar10_pytorch_const", "optimizations": {"aggregation_frequency": 1, "average_aggregated_gradients": true, "average_training_metrics": true, "gradient_compression": false, "grad_updates_size_file": null, "mixed_precision": "O0", "tensor_fusion_threshold": 64, "tensor_fusion_cycle_time": 5, "auto_tune_tensor_fusion": false}, "perform_initial_validation": false, "profiling": {"enabled": false, "begin_on_batch": 0, "end_after_batch": null, "sync_timings": true}, "project": "", "records_per_epoch": 50000, "reproducibility": {"experiment_seed": 1688150111}, "resources": {"max_slots": null, "slots_per_trial": 1, "weight": 1, "native_parallel": false, "shm_size": null, "resource_pool": "default", "priority": null, "devices": []}, "scheduling_unit": 100, "searcher": {"max_length": {"epochs": 4}, "metric": "validation_error", "name": "single", "smaller_is_better": true, "source_checkpoint_uuid": null, "source_trial_id": null}, "workspace": "", "slurm": {}, "pbs": {}}
<info> [2023-06-30 18:35:20] [88c6cd33] [7074] root: detected 2 rocm gpus
<info> [2023-06-30 18:35:20] [88c6cd33] [7074] root: Running task container on agent_id=o186i122, hostname= with visible GPUs ['0xcd16cf15a61d2854', '0xdcd8a4ff9ca93cfa']
...
```

Checked for regression when using `slot-type=cuda`.

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: agent/pkg/singularity/singularity.go [hunk 1]
```diff
@@ -387,6 +378,49 @@ func (s *SingularityClient) RunContainer(
...
+		cudaVisibleDevicesVar := strings.Join(cudaVisibleDevices, ",")
+		cmd.Env = append(cmd.Env,
+			fmt.Sprintf("SINGULARITYENV_CUDA_VISIBLE_DEVICES=%s", cudaVisibleDevicesVar),
+			fmt.Sprintf("APPTAINERENV_CUDA_VISIBLE_DEVICES=%s", cudaVisibleDevicesVar),
+		)
+	}
+	if req.DeviceType == device.ROCM {
+		// Avoid this problem: https://github.com/determined-ai/determined-ee/pull/922
+		// by not setting both ROCR_VISIBLE_DEVICES & CUDA_VISIBLE_DEVICES
+		s.addToCommandEnvironment(cmd, "ROCR_VISIBLE_DEVICES")
```
Comment (jerryharrow) [inline agent/pkg/singularity/singularity.go:399]: Why are we not doing the same as on 392-393 -- i.e. 
SINGULARITYENV_ROCR_VISIBLE_DEVICES & APPTAINERENV_ROCR_VISIBLE_DEVICES?

Or switch these to use  s.addToCommandEnvironment(cmd,...)?

Seems like they all should be treated the same.

File: agent/pkg/singularity/singularity.go [hunk 2]
```diff
@@ -387,6 +378,49 @@ func (s *SingularityClient) RunContainer(
...
+	cmd.Env = append(cmd.Env, fmt.Sprintf("PATH=%s", os.Getenv("PATH")))
+
+	s.addToCommandEnvironment(cmd, "http_proxy")
+	s.addToCommandEnvironment(cmd, "https_proxy")
+	s.addToCommandEnvironment(cmd, "no_proxy")
+	s.addToCommandEnvironment(cmd, "APPTAINER_CACHEDIR")
+	s.addToCommandEnvironment(cmd, "SINGULARITY_CACHEDIR")
+}
+
+func (s *SingularityClient) addToCommandEnvironment(cmd *exec.Cmd, variable string) {
```
Comment (jerryharrow) [inline agent/pkg/singularity/singularity.go:417]: The prior name include ifSet suffix -- it still only adds to the environment if set, so seems better to include that in the name.

File: agent/pkg/singularity/singularity.go (hunk 1 above)
Comment (phillip-gaisford) [reply agent/pkg/singularity/singularity.go:399]: Those are not variables that Singularity [documents](https://docs.sylabs.io/guides/latest/user-guide/gpu.html#) as being supported.

File: agent/pkg/singularity/singularity.go (hunk 2 above)
Comment (phillip-gaisford) [reply agent/pkg/singularity/singularity.go:417]: OK

File: agent/pkg/singularity/singularity.go (hunk 1 above)
Comment (jerryharrow) [reply agent/pkg/singularity/singularity.go]: OK.  Probably a comment indicating that --nv & --rocm support in Singularity/Apptainer requires special treatment of the variables.

File: agent/pkg/singularity/singularity.go [hunk 3]
```diff
@@ -387,6 +378,51 @@ func (s *SingularityClient) RunContainer(
...
+		cudaVisibleDevicesVar := strings.Join(cudaVisibleDevices, ",")
+		cmd.Env = append(cmd.Env,
+			fmt.Sprintf("SINGULARITYENV_CUDA_VISIBLE_DEVICES=%s", cudaVisibleDevicesVar),
+			fmt.Sprintf("APPTAINERENV_CUDA_VISIBLE_DEVICES=%s", cudaVisibleDevicesVar),
+		)
+	}
+	if req.DeviceType == device.ROCM {
+		// Avoid this problem: https://github.com/determined-ai/determined-ee/pull/922
+		// by not setting both ROCR_VISIBLE_DEVICES & CUDA_VISIBLE_DEVICES
+		s.addToTargetCommandEnvironmentIfSet(cmd, "ROCR_VISIBLE_DEVICES")
```
Comment (jerryharrow) [inline agent/pkg/singularity/singularity.go:399]: Interesting that it doesn't need to be 

SINGULARITYENV_ROCR_VISIBLE_DEVICES

???

File: agent/pkg/singularity/singularity.go (hunk 1 above)
Comment (phillip-gaisford) [reply agent/pkg/singularity/singularity.go]: Bradley has a story to move us away from `--nv`, so likely this will be revised soon.

**Review by jerryharrow — APPROVED** [review]
